rules/Multifactor-Google-Authenticator-Do-Not-Rename.js

```diff
@@ -1,35 +1,74 @@
 function (user, context, callback) {
-
-    var mfa_enabled_connection = [
+    var AUTHENTICATOR_LABEL = 'MOJ Analytical Platform (Alpha)';
+    var DISABLED_CLIENTS = [
+        'oUb1V330oXKyMpTagAYDzWDY10U4ffWF', // kubectl-oidc
+    ];
+    var ENABLED_CONNECTIONS = [
         'github',
-        'google-oauth2'
-    ].indexOf(context.connection) !== -1;
+        'google-oauth2',
+    ];
+    var MFA_CHALLENGE_EVERY_MINUTES = 8 * 60; // 8 hours
 
-    // Exclude the following clients from mfa
-    var mfa_disabled_clients = [
-        'oUb1V330oXKyMpTagAYDzWDY10U4ffWF'
-    ].indexOf(context.clientID) === -1;
 
-    var user_with_mfa = user.app_metadata && user.app_metadata.use_mfa;
+    var enabled_for_user = !user.app_metadata || user.app_metadata.use_mfa !== false;
+    var enabled_for_connection = ENABLED_CONNECTIONS.indexOf(context.connection) !== -1;
+    var enabled_for_client = DISABLED_CLIENTS.indexOf(context.clientID) === -1;
 
-    if (
-        mfa_enabled_connection &&
-        mfa_disabled_clients
-    ) {
+    if ( !enabled_for_user || !enabled_for_connection || !enabled_for_client ) {
+        console.log('2FA disabled for user, connection or client');
+        return callback(null, user, context);
+    }
 
-        context.multifactor = {
-            provider: 'google-authenticator',
+    getUserProfile(user)
+        .then(function (user_profile) {
+            if ( !loggedInRecently(user_profile) ) {
+                console.log('user did not log in recently - enabling MFA');
+                enableMFA(context);
+            }
 
-            // optional, the label shown in the authenticator app
-            issuer: 'MOJ Analytical Platform (Alpha)',
+            return callback(null, user, context);
+        })
+        .catch(function(err) {
+            enableMFA(context);
+            return callback(null, user, context);
+        });
 
-            // optional, the key to use for TOTP. By default one is generated for you
-            // key: '{YOUR_KEY_HERE}',
 
+    // Utility functions
+
+    function enableMFA(context) {
+        context.multifactor = {
+            provider: 'google-authenticator',
+            // optional, the label shown in the authenticator app
+            issuer: AUTHENTICATOR_LABEL,
             // optional, defaults to true. false forces 2FA every time.
-            allowRememberBrowser: false
+            allowRememberBrowser: false,
         };
     }
 
-    callback(null, user, context);
+    function loggedInRecently(user_profile) {
+        if (user_profile.last_login) {
+            var secs_from_last_login = (new Date() - new Date(user_profile.last_login)) / 1000;
+
+            if (secs_from_last_login < 5) {
```

Comment:
This this the fix for the problem you were having? I think a comment might be useful here (😱)